### PR TITLE
docs: add section on HttpErrors in Controllers

### PR DESCRIPTION
**Pull request by shimks**
## Description
- add blurb on HttpErrors under Controllers
- connected to https://github.com/strongloop/loopback-next/issues/551

### CODE REVIEW

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [inline pages/en/lb4/Controllers.md]: I'd mention that we're actually re-exporting the `http-errors` package: https://www.npmjs.com/package/http-errors

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [inline pages/en/lb4/Controllers.md]: ~Whoa! I don't think I've ever used 412 in my life!
...I like it.~

EDIT: Sorry, now that I looked it up, it's not meant to be used in this way.

From [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/412):
> The HTTP 412 Precondition Failed client error response code indicates that access to the target resource has been denied. This happens with conditional requests on methods other than GET or HEAD when the condition defined by the If-Unmodified-Since or If-None-Match headers is not fulfilled. In that case, the request, usually an upload or a modification of a resource, cannot be made and this error response is sent back.

**Review by kjdelisle — APPROVED** [review]
One minor thing you should add, but I'll approve it now. Excellent example!

**Review by kjdelisle — CHANGES_REQUESTED** [review]
I just had to Google it, didn't I?
I was curious about how code 412 worked and it's not actually intended to be used this way.

File: pages/en/lb4/Controllers.md
Comment (jannyHou) [inline pages/en/lb4/Controllers.md]: I think the `...` here is not a necessary, same to the other 4 `...`

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [reply pages/en/lb4/Controllers.md]: I'd change it to be a 422 Unprocessable Entity. The syntax is correct, but the _semantics_ of a request whose limit is a non-positive integer are incorrect, so it'd be the best match, IMO.

That being said, there's still debate about whether or not people should use 400 for that purpose, so either one is cool with me.

**Comment by kjdelisle** [comment]
@shimks If you wanted to dive into the murky depths of "which code do I use", feel free to start here! https://stackoverflow.com/questions/16133923/400-vs-422-response-to-post-of-data

File: pages/en/lb4/Controllers.md
Comment (jannyHou) [inline pages/en/lb4/Controllers.md]: `this.repository` would be the correct property name.
Each controller has a property `repository`

File: pages/en/lb4/Controllers.md
Comment (b-admike) [inline pages/en/lb4/Controllers.md]: You can mention here if `HttpErrors` has an exhaustive list of http errors or a select ones which are common. Also, you can mention that it handles the status codes and messages for the errors you want, etc.

File: pages/en/lb4/Controllers.md
Comment (b-admike) [inline pages/en/lb4/Controllers.md]: why are we throwing a new error here? shouldn't 
```
try { 
controller.list(0);
} catch (err) {
...
}
```
be enough?

File: pages/en/lb4/Controllers.md
Comment (jannyHou) [inline pages/en/lb4/Controllers.md]: ~please specify the type of `limit` instead of assigning a value :) (or maybe I misunderstand the code here....)~

my bad, I do miss an important thing(define default value in ts), the code is correct and elegant here!

File: pages/en/lb4/Controllers.md
Comment (jannyHou) [inline pages/en/lb4/Controllers.md]: nitpick: I think in javascript `number` could be integer or float, so any number between 0~1 is still positive

> As in JavaScript, all numbers in TypeScript are floating point values.

https://www.typescriptlang.org/docs/handbook/basic-types.html

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [reply pages/en/lb4/Controllers.md]: Good catch! Sorry I missed that. Actually, it's even worse because the catch block will catch it instead.

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [inline pages/en/lb4/Controllers.md]: method should be async

**Review by jannyHou — COMMENTED** [review]
nice doc! a few comments :)

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [inline pages/en/lb4/Controllers.md]: function should return a rejected promise instead
`return Promise.resolve(new HttpErrors.UnprocessableEntity('limit is non-positive'));`

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [inline pages/en/lb4/Controllers.md]: `return await this.messages.find({limit});`

**Review by kjdelisle — CHANGES_REQUESTED** [review]

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [reply pages/en/lb4/Controllers.md]: and it needs a bracket

File: pages/en/lb4/Controllers.md
Comment (virkt25) [inline pages/en/lb4/Controllers.md]: I think most of the docs use `ts` not `js` ... switching to which would require you to add types for variables as per @jannyHou 's comment.

File: pages/en/lb4/Controllers.md
Comment (virkt25) [reply pages/en/lb4/Controllers.md]: +1 Would be helpful to have a list of all error's available (or atleast the common ones and a link to all the options in `http-errors` module).

File: pages/en/lb4/Controllers.md
Comment (shimks) [reply pages/en/lb4/Controllers.md]: what happens if controller.list(0) doesn't throw error as expected? Shouldn't expects() in the catch block not run and cause the test to pass when it shouldn't?

Comment (virkt25) [reply pages/en/lb4/Controllers.md]: Move `expect` statements to be outside of the try/catch block and you can assign `err` to another variable in the catch block so it's accessible outside. 

```
try {
  controller.list(0)
} catch (err) {
  goterr = err
}
expect(goterr).to.exist();
// ... 
```

File: pages/en/lb4/Controllers.md
Comment (jannyHou) [reply pages/en/lb4/Controllers.md]: @virkt25 yesterday I learned why we use `js` not `ts` lol... `.ts` doesn't recognize some keyword like `async`:

```ts
// a .ts file
async
```

```js
// a .js file
async
```

Comment (virkt25) [reply pages/en/lb4/Controllers.md]: :o interesting. I didn’t know that. Thanks for enlightening me @jannyHou

File: pages/en/lb4/Controllers.md
Comment (bajtos) [reply pages/en/lb4/Controllers.md]: I am personally fine with this approach. If we switch the order of the two `expect()` lines below and check the message first, then we will get a nice test failure message too - something along the lines "expected /precondition failed/" but got "should have thrown an error".

File: pages/en/lb4/Controllers.md
Comment (bajtos) [reply pages/en/lb4/Controllers.md]: Controllers should use dependency injection to receive a Repository instance. Let's simplify this example and make the controller "pure" with no dependencies.

**Review by bajtos — COMMENTED** [review]
Could you please edit [Thinking in LoopBack] too and add a link referencing your new content? Here is the section in Thinking in LoopBack that is describing how to report errors from controller methods: http://loopback.io/doc/en/lb4/Thinking-in-LoopBack.html#handle-product-not-found-error

File: pages/en/lb4/Controllers.md
Comment (bajtos) [reply pages/en/lb4/Controllers.md]: ❗️ I am pretty certain this test will be confusing to many other readers too. Regardless of the  approach we decide to follow, there should be code comment(s) explaining why and how this test works, why it's important to be prepared for the case when the controller method does not throw any error, and how we handled that.

File: pages/en/lb4/Controllers.md
Comment (shimks) [reply pages/en/lb4/Controllers.md]: Could you clarify what you mean by that @bajtos ? Is it right for me to assume that I should continue to not use dependency injection here?

File: pages/en/lb4/Controllers.md
Comment (dhmlau) [reply pages/en/lb4/Controllers.md]: It would be better if we put some content instead of the `...`.  e.g. add your other imports here, or add your logics here.

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [inline pages/en/lb4/Controllers.md]: `await controller.list(0.4);`

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [inline pages/en/lb4/Controllers.md]: `it('returns 422 Unprocessable Entity for non-positive limit', async () => {`

**Comment by shimks** [comment]
I should mention, the code was based on http://loopback.io/doc/en/lb4/Controllers.html#writing-controller-methods. Since I've built on top of that section and made changes, should I also fix that section up a little?

**Review by kjdelisle — CHANGES_REQUESTED** [review]
Couple more minor nits, but looking good!

File: pages/en/lb4/Controllers.md
Comment (kjdelisle) [inline pages/en/lb4/Controllers.md]: `if (!Number.isInteger(limit) || limit  < 1)`

Comment (kjdelisle) [reply pages/en/lb4/Controllers.md]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger

**Comment by kjdelisle** [comment]
>I should mention, the code was based on http://loopback.io/doc/en/lb4/Controllers.html#writing-controller-methods. Since I've built on top of that section and made changes, should I also fix that section up a little?

Yeah, I'd put your cleaned-up code there in its place once you've got the green light for the rest of it.
(Omit the tests for that example, though)

**Review by crandmck — APPROVED** [review]

**Review by b-admike — APPROVED** [review]
LGTM once all the remaining comments are addressed. Great job @shimks 👍

**Review by jannyHou — APPROVED** [review]

**Review by kjdelisle — APPROVED** [review]

File: pages/en/lb4/Controllers.md
Comment (bajtos) [inline pages/en/lb4/Controllers.md]: please use `let` instead of `var`, see http://loopback.io/doc/en/contrib/style-guide.html#variable-declarations

File: pages/en/lb4/Controllers.md
Comment (bajtos) [inline pages/en/lb4/Controllers.md]: This part has two problems:
 - It calls `Promise.resolve()` instead of `Promise.reject()`, therefore the HTTP error won't be treated as an error, it won't be passed to `catch` handlers.
 - The Promise is not returned (missing `return` statement). As a result, this promise is ignored, and because it was created via `Promise.resolve`, no warning is printed.

Please simplify this part by using `throw`. The mechanism powering `async` function will convert the thrown error to a rejected promise automatically for you.

```ts
if (!Number.isInteger(limit) || limit < 1) {
  throw new HttpErrors.UnprocessableEntity('limit is non-positive'));
}
```

Also please copy your test and your implementation code into a small application and execute them to make sure both snippets work as expected!

**Review by bajtos — CHANGES_REQUESTED** [review]

**Review by b-admike — COMMENTED** [review]
Looks like you'll need to rebase your branch:
```
[workspace] $ /bin/sh /tmp/jenkins6528781950920441340.sh
**************************************************
**
**  Checking history of add-HttpErrors against gh-pages
**
**  PASS: add-HttpErrors merge-base and fork-point match, good job
**  FAIL: add-HttpErrors is behind gh-pages by 7 commits, rebase needed
**
**************************************************
Build step 'Execute shell' marked build as failure
```

**Review by bajtos — APPROVED** [review]
👍

**Pull request merged**
